hermes/bin/src/runtime_extensions/hermes/cardano/mod.rs

```diff
@@ -1,7 +1,788 @@
+struct ChainFollowerHandle {
+    /// Commands channel sender.
+    cmd_tx: ChainFollowerHandleCommandSender,
+}
+
+impl ChainFollowerHandle {
+    /// Sends a command to the chain follower executor task to set its
+    /// read pointer to the given point.
+    fn set_read_pointer_sync(
+        &self, at: cardano_chain_follower::PointOrTip,
+    ) -> Result<Option<cardano_chain_follower::Point>> {
+        let (res_tx, res_rx) = tokio::sync::oneshot::channel();
+
+        self.cmd_tx
+            .blocking_send(ChainFollowerCommand::SetReadPointer(at, res_tx))
+            .map_err(|_| Error::InternalError)?;
+
+        res_rx
+            .blocking_recv()
+            .map_err(|_| Error::InternalError)?
+            .map_err(|_| Error::InternalError)
+    }
+
+    /// Sends a command to the chain follower executor task to stop following.
+    /// The follower continues active and following can be resumed by calling
+    /// [`Self::resume`].
+    fn stop(&self) -> Result<()> {
+        let (res_tx, res_rx) = tokio::sync::oneshot::channel();
+
+        self.cmd_tx
+            .blocking_send(ChainFollowerCommand::Stop(res_tx))
+            .map_err(|_| Error::InternalError)?;
+
+        drop(res_rx.blocking_recv());
+
+        Ok(())
+    }
+
+    /// Sends a command to the chain follower executor task to resume following
+    /// from the point it was previously stopped.
+    ///
+    /// This has no effect if the follower is not stopped.
+    fn resume(&self) -> Result<()> {
+        let (res_tx, res_rx) = tokio::sync::oneshot::channel();
+
+        self.cmd_tx
+            .blocking_send(ChainFollowerCommand::Continue(res_tx))
+            .map_err(|_| Error::InternalError)?;
+
+        drop(res_rx.blocking_recv());
+
+        Ok(())
+    }
+}
+
+/// Hermes application module subscription state.
+#[derive(Default)]
+struct SubscriptionState {
+    /// Whether the module is subscribed to receive block events.
+    subscribed_to_blocks: bool,
+    /// Whether the module is subscribed to receive transaction events.
+    subscribed_to_txns: bool,
+    /// Whether the module is subscribed to receive rollback events.
+    subscribed_to_rollbacks: bool,
+    /// Handle to the cardano chain follower from which the module is receiving
+    /// events.
+    follower_handle: Option<ChainFollowerHandle>,
+    ///
+    current_slot: u64,
+}
+
+/// Cardano Runtime Extension state.
+struct State {
+    /// Handle to the Tokio runtime background thread.
+    tokio_rt_handle: TokioRuntimeHandle,
+    /// Mapping of application module subscription states.
+    subscriptions:
+        DashMap<(HermesAppName, ModuleId, cardano_chain_follower::Network), SubscriptionState>,
+    /// Chain followers configured only for reading blocks.
+    readers: DashMap<cardano_chain_follower::Network, cardano_chain_follower::Follower>,
+}
+
+/// Cardano Runtime Extension internal state.
+static STATE: once_cell::sync::Lazy<State> = once_cell::sync::Lazy::new(|| {
```

Comment:
What do you think to move initialisation as a part of the `HermesReactor` and place a thread `JoinHandle` as a field there, so it reactor will maintain an execution loop of the follower.
The same as it was done for hermes event queue

Reply:
I think holding the thread join handle somewhere is a good ideia. But I think it would be better that we had some function that the reactor would call when it's time for the runtime extension to be shutdown instead of leaking the logic of runtime extensions on the reactor. 

I mean, having the logic for the queue in the reactor makes sense because the event queue is managed by the reactor but I don't think the reactor should have code for every background thread we may want to spawn.

WDYT?

Reply:
I'm thinking along the same lines for the cron queue.

Reply:
Yeah, I agree that we should gracefully shutdown Hermes. But I think the reactor should signal the Runtime Extensions that they should shutdown. Just waiting on a `JoinHandle` is not enough for that.

In the case the background thread is running a loop (like the Cardano Runtime Extension) there needs to be some kind of communication (or something like dropping channel receivers/senders) in order for the loop to be stopped. Otherwise the reactor would wait forever.